File: MusicRemoteController/PlayerModeViewController.m
//
//  PlayerModeViewController.m
//  MusicRemoteController
//
//  Created by Yusuke Sato on 2014/03/24.
//  UPDATE VERSION 1.2.0 2015/07/26.
//  Copyright (c) 2014年 Yusuke Sato. All rights reserved.
//

#import "PlayerModeViewController.h"

#define SERVICE_UUID_STRING @"7865087B-D9D0-423A-9C80-042D9BBEA524"
#define MUSICINFO_UUID_STRING @"085BACFC-5FCD-49D1-B7D7-A331901F6DDE"
#define BUTTON_UUID_STRING @"608072DD-6825-4293-B3E7-324CF0B5CA08"

@interface PlayerModeViewController ()
@property CBUUID *uuid;
@end

@implementation PlayerModeViewController
-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:NO];
    
    // MPMediaPlayer
    self.player = [MPMusicPlayerController iPodMusicPlayer];
    
    if ([self.player playbackState] == MPMusicPlaybackStateStopped){
        [self.player setQueueWithQuery:[MPMediaQuery songsQuery]];
    }
    
    self.barVolume.backgroundColor = [UIColor clearColor];
    MPVolumeView *volumeView = [[MPVolumeView alloc] initWithFrame:self.barVolume.bounds];
    [self.barVolume addSubview:volumeView];
    
    self.ncenter = [NSNotificationCenter defaultCenter];
    [self.ncenter addObserver:self selector:@selector(playbackStateChange) name:MPMusicPlayerControllerPlaybackStateDidChangeNotification object:self.player];
    [self.ncenter addObserver:self selector:@selector(nowPlayingItemChanged) name:MPMusicPlayerControllerNowPlayingItemDidChangeNotification object:self.player];
    [self.ncenter addObserver:self selector:@selector(volumeChange) name:MPMusicPlayerControllerVolumeDidChangeNotification object:self.player];
    [self.player beginGeneratingPlaybackNotifications];
    
    
    // CoreBluetooth
    self.serviceUUID = [CBUUID UUIDWithString:SERVICE_UUID_STRING];
    self.buttonUUID = [CBUUID UUIDWithString:BUTTON_UUID_STRING];
    self.musicInfoUUID = [CBUUID UUIDWithString:MUSICINFO_UUID_STRING];
    self.peripheralManager = [[CBPeripheralManager alloc] initWithDelegate:self queue:nil];
}

//==========================
// CoreBluetooth Programing
//==========================
-(void) peripheralManagerDidUpdateState:(CBPeripheralManager *)peripheral
{
    NSLog(@"peripheralManagerDidUpdateState");
    if (peripheral.state == CBPeripheralManagerStatePoweredOn){
        
        self.buttonCharacteristic = [[CBMutableCharacteristic alloc] initWithType:self.buttonUUID properties:(CBCharacteristicPropertyRead | CBCharacteristicPropertyWrite | CBCharacteristicPropertyNotify) value:nil permissions: (CBAttributePermissionsReadEncryptionRequired | CBAttributePermissionsWriteEncryptionRequired)];

        self.musicInfoCharacteristic = [[CBMutableCharacteristic alloc] initWithType:self.musicInfoUUID properties:(CBCharacteristicPropertyNotify | CBCharacteristicPropertyRead) value:nil permissions: (CBAttributePermissionsReadEncryptionRequired)];
        
        CBMutableService *service = [[CBMutableService alloc] initWithType:self.serviceUUID primary:YES];
        [service setCharacteristics:@[self.buttonCharacteristic,self.musicInfoCharacteristic]];
        [peripheral addService:service];
        
    }
}

-(void) peripheralManager:(CBPeripheralManager *)peripheral didAddService:(CBService *)service error:(NSError *)error
{
    NSLog(@"didAddService");
    if (error){
        NSLog(@"SERVICE ADD ERROR");
        NSLog(@"%@",[error localizedDescription]);
    }else{
        [peripheral startAdvertising:@{CBAdvertisementDataLocalNameKey:@"MRC", CBAdvertisementDataServiceUUIDsKey:@[self.serviceUUID]}];
    }
}

- (void)peripheralManager:(CBPeripheralManager *)peripheral didReceiveWriteRequests:(NSArray *)requests {
    
    for (CBATTRequest *aRequest in requests) {
        
        if ([aRequest.characteristic.UUID isEqual:self.buttonUUID]) {
            
            // CBCharacteristicのvalueに、CBATTRequestのvalueをセット
            [self pushRemoteController:aRequest.value];
        }
    }
    
    // リクエストに応答
    [self.peripheralManager respondToRequest:requests[0]
                                  withResult:CBATTErrorSuccess];
}

- (void)peripheralManager:(CBPeripheralManager *)peripheral
    didReceiveReadRequest:(CBATTRequest *)request
{
    if ([request.characteristic.UUID isEqual:self.buttonUUID]) {
        
        // CBMutableCharacteristicのvalueをCBATTRequestのvalueにセット
        
        request.value = [self getResponseValue];
        
        // リクエストに応答
        [self.peripheralManager respondToRequest:request
                                      withResult:CBATTErrorSuccess];
    }
    
    if ([request.characteristic.UUID isEqual:self.musicInfoUUID]){
        NSLog(@"Request Music Info");
        
        if (request.offset > self.musicInfoCharacteristic.value.length){
            NSLog(@"A");
            [self.peripheralManager respondToRequest:request withResult:CBATTErrorInvalidOffset];
            return;
        }else{
            NSLog(@"B");
            request.value = [self.musicInfoCharacteristic.value subdataWithRange:NSMakeRange(request.offset, self.musicInfoCharacteristic.value.length)];
            [self.peripheralManager respondToRequest:request withResult:CBATTErrorSuccess];
        }
        
        
        
    }
}

- (void) peripheralManager:(CBPeripheralManager *)peripheral
                         central:(CBCentral *)central
    didSubscribeToCharacteristic:(CBCharacteristic *)characteristic
{
    NSLog(@"Subscribeリクエストを受信");
}

-(void)pushRemoteController:(NSData *)val
{
    NSLog(@"REQUEST VALUE IS %@",val);
    int param = *(int *)([val bytes]);
    switch (param) {
        case 0 :
            if (_isPlaying == YES){
                [_player pause];
                _isPlaying = NO;
            }else{
                [_player play];
                _isPlaying = YES;
            }
            break;
        case 1 :
            NSLog(@"PeripheralでRevボタンが押されました。");
            [_player skipToPreviousItem];
            break;
        case 2: //早送り
            NSLog(@"PeripheralでNextボタンが押されました。");
            [_player skipToNextItem];
            break;
        case 3: // VolUp
            [self setVolume:1];
            break;
        case 4: // VolDown
            [self setVolume:0];
            break;
        default:
            NSLog(@"error");
            break;
    }
}
//==========================
// MediaPlayer Programing
//==========================
- (void) playerStateCheck
{
    int state = [_player playbackState];
    NSLog(@"%d",state);
    if (state == 1){
        [self.btnPlay setImage:[UIImage imageNamed:@"stop.png"] forState:UIControlStateNormal];
    }else{
        [self.btnPlay setImage:[UIImage imageNamed:@"Play.png"] forState:UIControlStateNormal];
    }
}


- (void) playbackStateChange
{
    NSLog(@"Playback State Item Change");
    [self playerStateCheck];
    
}

- (void) nowPlayingItemChanged
{
    NSLog(@"Playing Item Changed");
    MPMediaItem *NowPlayingItem = [self.player nowPlayingItem];
    NSString *SongTitle = [NowPlayingItem valueForProperty:MPMediaItemPropertyTitle];
    NSString *SongArtist = [NowPlayingItem valueForProperty:MPMediaItemPropertyArtist];
    MPMediaItemArtwork *SongArtwork = [NowPlayingItem valueForProperty:MPMediaItemPropertyArtwork];
    
    //====
    //再生中の音楽情報セット
    //====
    
    // タイトル
    self.lblTitle.text = SongTitle;
    
    // アーティスト
    self.lblArtist.text = SongArtist;
    
    // アートワーク
    UIImage *image = [SongArtwork imageWithSize:CGSizeMake(200.0f,200.0f)];
    if (image == NULL){
        image = [UIImage imageNamed:@"nonimage.png"];
    }
    [self.imageArtwork setImage:image];
    
    if (self.musicInfoCharacteristic != nil){
        
        NSMutableDictionary *dic = [NSMutableDictionary dictionary];
        [dic setValue:SongTitle forKey:@"TITLE"];
        [dic setValue:SongArtist forKey:@"ARTIST"];
        //[dic setObject:image forKey:@"ARTWORK"];
   
        NSData *updValue = [NSKeyedArchiver archivedDataWithRootObject:dic];
   
        
        [self.peripheralManager updateValue:updValue forCharacteristic:_musicInfoCharacteristic onSubscribedCentrals:nil];
    }
}

- (void) volumeChange
{
    NSLog(@"Volume Change");
    NSLog(@"%f",[self.player volume]);
}


- (IBAction)btnPrevPush:(id)sender {
    [self.player skipToPreviousItem];
}

- (IBAction)btnPlayPush:(id)sender {
    if(self.isPlaying == NO) {
        [self.btnPlay setImage:[UIImage imageNamed:@"stop.png"] forState:UIControlStateNormal];
        [self.player play];
        self.isPlaying = YES;
    }else{
        self.isPlaying = NO;
        [self.btnPlay setImage:[UIImage imageNamed:@"Play.png"] forState:UIControlStateNormal];
        [self.player pause];
        
    }
}

- (IBAction)btnNextPush:(id)sender {
    [_player skipToNextItem];
}

- (void) setVolume:(int)flg
{
    float vol = [_player volume];
    switch (flg) {
        case 1:
            vol += 0.0625;
            break;
        case 0:
            vol -= 0.0625;
            break;
    }
    
    if (vol <= 0.0 ) {
        vol = 0.0;
    }else if (vol >= 1.0){
        vol = 1.0;
    }
    [self.player setVolume:vol];
}

- (NSData *)getResponseValue{
    
    uint val;
    
    if (self.isPlaying){
        val = 1;
    }else{
        val = 0;
    }
    
    return [NSData dataWithBytes:&val length:1];
    
}

@end

File: MusicRemoteController/PlayerModeViewController.swift
//
//  PlayerModeViewController.swift
//  MusicRemoteController
//
//  Created by Yusuke Sato on 2017/09/30.
//  Copyright © 2017年 Yusuke Sato. All rights reserved.
//

import UIKit
import CoreBluetooth
import MediaPlayer

class PlayerModeViewController : NendViewController ,CBPeripheralManagerDelegate,UIAlertViewDelegate {

    @IBOutlet var imgArtwork: UIImageView!
    @IBOutlet var lblTitle: UILabel!
    @IBOutlet var lblArtist: UILabel!
    
    @IBOutlet var barSlider: UISlider!
    @IBOutlet var lblTime: UILabel!
    @IBOutlet var lblDuration: UILabel!
    
    @IBOutlet var btnPrev: UIButton!
    @IBOutlet var btnPlay: UIButton!
    @IBOutlet var btnNext: UIButton!
    
    var _timer : Timer!
    
    /** 通知情報 */
    var _playInfo : Data!
    
    /** ペリフェラルマネージャ */
    var _peripheralManager : CBPeripheralManager!
    /** Characteristic(ボタン)  */
    var _buttonCharacteristic : CBCharacteristic!
    /** Characteristic(音楽) */
    var _musicInfoCharacteristic: CBCharacteristic!
    /** Characteristic(通知) */
    var _notifyCharacteristic: CBCharacteristic!
    // プレーヤーインスタンス
    var _player : MPMusicPlayerController!
    // 通知センターインスタンス
    var _notificationCenter : NotificationCenter!
    
    let options : Dictionary = [
            CBCentralManagerOptionRestoreIdentifierKey: "MRC",
            CBCentralManagerOptionShowPowerAlertKey: true
        ] as [String : Any]
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if AuthUtil.isMpMediaLibrary(view: self) && AuthUtil.isCoreBluetoothAuth(view: self) {
            
            if (MPMediaQuery.songs().items?.count)! > 0 {
                // プレーヤーの初期化
                _player = MPMusicPlayerController.systemMusicPlayer
                // 通知センターの初期化
                _notificationCenter = NotificationCenter.default
                
                // 再生状態変更通知
                _notificationCenter.addObserver(self, selector: #selector(self.playbackStateChange), name: .MPMusicPlayerControllerPlaybackStateDidChange, object: _player)
                
                // 再生状態変更通知
                _notificationCenter.addObserver(self, selector: #selector(self.playingItemChange), name: NSNotification.Name.MPMusicPlayerControllerNowPlayingItemDidChange, object: _player)
                
                // 通知開始
                _player.beginGeneratingPlaybackNotifications()
                
                // 再生中でない場合適当な全曲からランダム
                if _player.playbackState == .stopped {
                    self._player.setQueue(with: .songs())
                }
                
                playingItemChange()
                
                playbackStateChange()
                
                _peripheralManager = CBPeripheralManager(delegate: self, queue: nil, options: options)
                
            }else{
                let alert : UIAlertView = UIAlertView(title: "再生可能な曲がありません。", message: "メニューへ戻ります。", delegate: self, cancelButtonTitle: "OK")
                alert.show()
            }
        }else{
            let alert : UIAlertView = UIAlertView(title: "アプリの動作に必要な許可がありません。", message: "メニューへ戻ります。", delegate: self, cancelButtonTitle: "OK")
            alert.show()
        }
    }
    
    // ===========================
    // CoreBluetooth
    // ===========================
    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        print("peripheralManagerDidUpdateState")
        
        if _peripheralManager.state.rawValue == CBPeripheralManagerState.poweredOn.rawValue {
            
            // Characteristic(ボタン)の初期化
            _buttonCharacteristic = CBMutableCharacteristic(type: UUIDS.BUTTON, properties: [.write,.notifyEncryptionRequired], value: nil, permissions: [.writeEncryptionRequired])
            
            // Characteristic(音楽)の初期化
            _musicInfoCharacteristic = CBMutableCharacteristic(type: UUIDS.MUSIC_INFO, properties: [.read ], value: nil, permissions: [.readEncryptionRequired])
            
            // Characteristic(通知)の初期化
            _notifyCharacteristic = CBMutableCharacteristic(type: UUIDS.NOTIFY, properties: [.notify], value: nil, permissions: [.readable])
            
            // サービス作成
            let service : CBMutableService = CBMutableService(type: UUIDS.SERVICE, primary: true)
            
            // サービスにCharacteristics追加
            service.characteristics = [_buttonCharacteristic, _notifyCharacteristic, _musicInfoCharacteristic]
            
            // サービスを追加
            peripheral.add(service)
        }
    }

    /** サービス追加結果 */
    func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
        print("didAddService")
        
        if error != nil {
            print("ERROR: \(String(describing: error))")
        }else{
            peripheral .startAdvertising([CBAdvertisementDataLocalNameKey:"MRC", CBAdvertisementDataServiceUUIDsKey:[UUIDS.SERVICE]])
        }
    }
    
    /** アドバタイズ開始 */
    func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
        print("peripheralManagerDidStartAdvertising")
        
        if error != nil {
            print("ERROR: \(String(describing: error))")
        }
    }
    
    /** 書き込み要求 */
    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
        print("didReceiveWrite")
        
        for request in requests {
            if request.characteristic.uuid.isEqual(_buttonCharacteristic.uuid) {
                WriteRequest(data: requests[0].value!)
            }
        }
        _peripheralManager.respond(to: requests[0], withResult: .success)
    }
    
    /** 読み込み要求 */
    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
        
        // RECEIVED MUSIC INFO READ
        if (request.characteristic.uuid.isEqual(UUIDS.MUSIC_INFO)) {
            if request.offset > _playInfo.count {
                
            }else{
                
                print(Range(uncheckedBounds: (request.offset, _playInfo.count)))
                
                request.value = _playInfo.subdata(in: Range(uncheckedBounds: (request.offset, _playInfo.count)))
                
                peripheral.respond(to: request, withResult: .success)
                
                print("Read success")
            }
        }else{
            print("Read fail: wrong characteristic uuid:", request.characteristic.uuid)
        }
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didUnsubscribeFrom characteristic: CBCharacteristic) {
        print("didUnsubscribeFrom")
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, willRestoreState dict: [String : Any]) {
        print("willRestoreState")
        
        let val = dict[CBCentralManagerRestoredStatePeripheralsKey]
        
        if val == nil {
            self.peripheralManagerDidUpdateState(peripheral)
        }
    }
    
    // ===========================
    // Event
    // ===========================
    /** 再生状態変更 */
    @objc func playbackStateChange() {
        print("playbackStateChange")
        
        if _player.playbackState == .playing {
            _timer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(self.updatePlayingTimer), userInfo: nil, repeats: true)
            btnPlay.setImage(UIImage(named: "stop.png"), for: .normal)
        }else{
            btnPlay.setImage(UIImage(named: "Play.png"), for: .normal)
        }
        
        if _buttonCharacteristic != nil{
            _peripheralManager.updateValue(Data(bytes: [UInt8(_player.playbackState.rawValue)]), for: _buttonCharacteristic as! CBMutableCharacteristic, onSubscribedCentrals: nil)
        }
    }
    
    /** 再生楽曲変更 */
    @objc func playingItemChange(){
        print("playingItemChange")
        
        setDisplay()
        
        setPlayInfo()
        
        if _notifyCharacteristic != nil {
            // 変更通知
            _peripheralManager.updateValue(_playInfo, for: _notifyCharacteristic as! CBMutableCharacteristic, onSubscribedCentrals: nil)
        }
    }
    
    /** currentTime更新 */
    @objc func updatePlayingTimer(){
        barSlider.value = Float(_player.currentPlaybackTime)
        lblTime.text = convertDoubleToTimeString(time: _player.currentPlaybackTime)
    }
    
    // ===========================
    // Action
    // ===========================
    /** アラートビューボタン押下 */
    func alertView(_ alertView: UIAlertView, clickedButtonAt buttonIndex: Int) {
        if buttonIndex == 0 {
            self.navigationController?.popViewController(animated: true)
        }
    }
    
    @IBAction func clickedPrev(_ sender: Any) {
        _player.skipToPreviousItem()
    }
    
    @IBAction func clickedPlay(_ sender: Any) {
        
        if _player.playbackState == .stopped || _player.playbackState == .paused {
            _player.play()
        }else if _player.playbackState == .playing{
            _player.pause()
        }
    }
    
    @IBAction func clickedNext(_ sender: Any) {
        _player.skipToNextItem()
    }
    
    @IBAction func changePosition(_ sender: Any) {
        _player.currentPlaybackTime = Double(barSlider.value)
        updatePlayingTimer()
    }
    
    // ===========================
    // Other
    // ===========================
    /** DOUBLE => hh:mm:ss */
    func convertDoubleToTimeString(time : Double) -> String{
        
        let s : Int = Int(time.truncatingRemainder(dividingBy: 60))
        let m : Int = Int(((time - Double(s)) / 60).truncatingRemainder(dividingBy: 60))
        let h : Int = Int(((time - Double(m) - Double(s)) / 3600 ).truncatingRemainder(dividingBy: 3600))
        
        return String(format: "%02d:%02d:%02d", h, m, s)
    }
    
    /** 画面表示 */
    func setDisplay(){
        var title : String = "";
        var artist : String = "";
        var artwork : UIImage = UIImage(named: "nonimage.png")!
        var durationTime : String = "00:00:00"
        var max : Float = Float(1)
        
        if (_player.nowPlayingItem != nil){
            let item : MPMediaItem = _player.nowPlayingItem!
            
            if item.title != nil {
                title = item.title!
            }else{
                title = ""
            }
            
            if item.artist != nil{
                artist = item.artist!
            }else{
                artist = ""
            }
            
            max = Float(item.playbackDuration)
            durationTime = convertDoubleToTimeString(time: (_player.nowPlayingItem?.playbackDuration)!)
            
            if (item.artwork != nil){
                artwork = (item.artwork?.image(at: CGSize(width: 200, height: 200)))!
            }
        }
        
        lblTitle.text = title;
        lblArtist.text = artist;
        imgArtwork.image = artwork
        lblDuration.text = durationTime
        barSlider.maximumValue = max
    }
    
    /** 通知情報設定 */
    func setPlayInfo(){
        var info : Dictionary<String, AnyObject> = [:]
        info["TITLE"] = lblTitle.text as AnyObject
        info["ARTIST"] = lblArtist.text as AnyObject
        
        do{
            _playInfo = try JSONSerialization.data(withJSONObject: info, options: .prettyPrinted)
        }catch{
            _playInfo = nil
        }
    }
    
    /** 書き込み要求 */
    func WriteRequest(data: Data){
        let val = data.hashValue
        
        if val == 0 {
            // Centralで再生ボタンが押された
            if _player.playbackState == .playing {
                // 再生中の場合、一時停止
                _player.pause()
            }else{
                _player.play()
            }
        }else if val == 1 {
            // CentralでNextボタンが押された
            _player.skipToNextItem()
        }else if val == 2 {
            // CentralでPrevボタンが押された
            _player.skipToPreviousItem()
        }else if val == 3 {
            // CentralでVolume Upが押された
            setVolume(mode: true)
        }else if val == 4 {
            // CentralでVolume Downが押された
            setVolume(mode: false)
        }
    }
    
    /** 音量設定 */
    func setVolume(mode: Bool){
        var vol : Double = _player.value(forKey: "volume") as! Double
        if (mode == false){
            vol -= 0.0625
            if (vol <= 0){
                vol = 0
            }
        }else{
            vol += 0.0625
            if (vol >= 1){
                vol = 1
            }
        }
        
        _player.setValue(vol, forKey: "volume")
    }
}
